src/librustc_typeck/collect.rs

```diff
@@ -1923,7 +1923,7 @@ fn get_or_create_type_parameter_def<'a,'tcx>(ccx: &CrateCtxt<'a,'tcx>,
                 param.id,
                 param.span,
                 format!("defaults for type parameters are only allowed on type definitions, \
-                         like `struct` or `enum`"));
```

Comment:
Seems like if we're going to try and be exhaustive, we ought to include `type` aliases too. I also think we should add `trait` into the list with `struct` and `enum`, like so:

    defaults for type parameters are only allowed in `struct`, `enum`, `type`, or `trait` definitions